src/lib/update-lock.ts

```diff
@@ -82,6 +82,75 @@ async function dispose(
 	}
 }
 
+/**
+ * Given a lockfile path `p`, return a tuple [appId, serviceName] of that path.
+ * Paths are assumed to end in the format /:appId/:serviceName/(resin-)updates.lock.
+ */
+function getIdentifiersFromPath(p: string) {
+	const parts = p.split('/');
+	if (parts.pop()?.match(/updates\.lock/) === null) {
+		return [];
+	}
+	const serviceName = parts.pop();
+	const appId = parts.pop();
+	return [appId, serviceName];
+}
+
+type LockedEntity = { appId: number; services: string[] };
+
+/**
+ * A map of locked services by appId.
+ * Exported for tests only; getServicesLockedByAppId is the public generator interface.
+ */
+export class LocksTakenMap extends Map<number, Set<string>> {
+	constructor(lockedEntities: LockedEntity[] = []) {
+		// Construct a Map<number, Set<string>> from user-friendly input args
+		super(
+			lockedEntities.map(({ appId, services }) => [appId, new Set(services)]),
+		);
+	}
+
+	// Add one or more locked services to an appId
+	public add(appId: number, services: string | string[]): void {
+		if (typeof services === 'string') {
+			services = [services];
+		}
+		if (this.has(appId)) {
+			const lockedSvcs = this.get(appId)!;
+			services.forEach((s) => lockedSvcs.add(s));
+		} else {
+			this.set(appId, new Set(services));
+		}
+	}
+
+	// Return an array copy of locked services under an appId
+	public getServices(appId: number): string[] {
+		return this.has(appId) ? Array.from(this.get(appId)!) : [];
+	}
+
+	// Return whether a service is locked under an appId
+	public isLocked(appId: number, service: string): boolean {
+		return this.has(appId) && this.get(appId)!.has(service);
+	}
+}
+
+/**
+ * Return a list of services that are locked by the Supervisor under each appId.
+ */
+export function getServicesLockedByAppId(): LocksTakenMap {
+	const locksTaken = lockfile.getLocksTaken();
+	const servicesByAppId = new LocksTakenMap();
+	for (const lockTakenPath of locksTaken) {
+		const [appId, serviceName] = getIdentifiersFromPath(lockTakenPath);
+		if (!StringIdentifier.is(appId) || !DockerName.is(serviceName)) {
+			continue;
+		}
+		const numAppId = +appId;
+		servicesByAppId.add(numAppId, serviceName);
```

Comment:
What if for some reason the folder only has `resin-updates.lock` and not `updates.lock`? Is the lock for the service taken in that case?

Reply:
If both or one of either file exists, the lock is considered taken by the Supervisor, as long as the locks taken are stored in memory. 

With the discussed change to 1) remove the in-memory store of lockfiles taken, 2) move to checking the filesystem, and 3) retain Supervisor-taken lockfiles between Supervisor restarts, locks would be considered taken by the Supervisor if one or both of `(resin-)updates.lock` exists in the filesystem with the `nobody` UID. These discussed changes will be added in a separate commit (will be linked here once complete). In the meantime I added a test case [here](https://github.com/balena-os/balena-supervisor/pull/2234/commits/c83b931d8643e9f842564ba81b074e44942c819e#diff-e7a2b32a616114c913967bbeb93301d1194c25d37b5c6d6393146a862d33b61fR339) which affirms that if only `resin-updates.lock` is present for some reason, the lock is still considered taken.

Reply:
I think it needs to consider the lock taken only if both files exist, as we don't know which file the container is using

Reply:
Added the `nobody` uid & both lockfiles present restrictions in this new commit: https://github.com/balena-os/balena-supervisor/commit/99ddf8387317b23bfec1e5aaa0161cbf7992ec51